csharp/Encryption/CryptoFactory.cs

```diff
@@ -0,0 +1,234 @@
+using System;
+using System.Runtime.InteropServices;
+using ParquetSharp.IO;
+
+namespace ParquetSharp.Encryption
+{
+    /// <summary>
+    /// Translates high-level encryption configuration into low-level encryption parameters
+    /// </summary>
+    public sealed class CryptoFactory : IDisposable
+    {
+        public delegate IKmsClient KmsClientFactory(ReadonlyKmsConnectionConfig config);
+
+        /// <summary>
+        /// Create a new CryptoFactory
+        /// </summary>
+        /// <param name="kmsClientFactory">Creates KMS clients from a connection configuration</param>
+        public unsafe CryptoFactory(KmsClientFactory kmsClientFactory)
+        {
+            var handle = ExceptionInfo.Return<IntPtr>(CryptoFactory_Create);
+            _handle = new ParquetHandle(handle, CryptoFactory_Free);
+
+            ExceptionInfo.Check(CryptoFactory_RegisterKmsClientFactory(
+                _handle.IntPtr,
+                CreateClientFactoryGcHandle(kmsClientFactory),
+                FreeGcHandleCallback,
+                CreateClientCallback,
+                WrapKeyCallback,
+                UnwrapKeyCallback));
+        }
+
+        /// <summary>
+        /// Get the encryption properties for a Parquet file.
+        /// If external key material is used then the path to the Parquet file must be provided.
+        /// </summary>
+        /// <param name="connectionConfig">The KMS connection configuration to use</param>
+        /// <param name="encryptionConfig">The encryption configuration to use</param>
+        /// <param name="filePath">The path to the Parquet file being written</param>
+        /// <returns>Encryption properties for the file</returns>
+        public FileEncryptionProperties GetFileEncryptionProperties(
+            KmsConnectionConfig connectionConfig,
+            EncryptionConfiguration encryptionConfig,
+            string? filePath = null)
+        {
+            var fileEncryptionPropertiesHandle = ExceptionInfo.Return<IntPtr, IntPtr, IntPtr, string?, IntPtr>(
+                _handle.IntPtr, connectionConfig.Handle.IntPtr, encryptionConfig.Handle.IntPtr, filePath, CryptoFactory_GetFileEncryptionProperties);
+            return new FileEncryptionProperties(fileEncryptionPropertiesHandle);
+        }
+
+
+        /// <summary>
+        /// Get decryption properties for a Parquet file.
+        /// If external key material is used then the path to the parquet file must be provided.
+        /// This CryptoFactory instance must remain alive and not disposed until after any files using these
+        /// decryption properties have been read, as internally the FileDecryptionProperties contains references to
+        /// data in the CryptoFactory that cannot be managed by ParquetSharp.
+        /// Failure to do so may result in native memory access violations and crashes that cannot be caught as exceptions.
```

Comment:
This is a bit awkward. I couldn't find a way to handle this nicely without exposing users to possible segfaults, as the `FileDecryptionProperties` class in C++ has a private constructor and can't be subclassed, so we can't do anything like combine it with a reference to the CryptoFactory. I'm hoping this can be fixed in Arrow so we can fix this limitation for the next version though.

Reply:
I've made a PR in Arrow to address this at https://github.com/apache/arrow/pull/40329

Reply:
That's been merged, so we should be able to remove this warning in the 16.0.0 release